catalog/dags/data_refresh/create_filtered_index_dag.py

```diff
@@ -0,0 +1,224 @@
+This parameter is only for use by the data refresh DAG and should not be
+used when manually triggering the DAG unless you are absolutely certain
+of what you are doing.
+"""
+import uuid
+from datetime import datetime, timedelta
+
+from airflow import DAG
+from airflow.decorators import task
+from airflow.models.param import Param
+from airflow.sensors.external_task import ExternalTaskSensor
+
+from common import ingestion_server
+from common.constants import DAG_DEFAULT_ARGS, XCOM_PULL_TEMPLATE
+from common.sensors.utils import get_most_recent_dag_run
+from data_refresh.data_refresh_types import DATA_REFRESH_CONFIGS
+
+
+# Note: We can't use the TaskFlow `@dag` DAG factory decorator
+# here because there's no way to set the media type as a static
+# variable in the context of the DAG because all arguments
+# passed to the DAG factory decorated by `@dag` are thrown
+# away: `@dag` decorated functions' arguments are treated
+# merely as DAG param aliases. We could use `@dag` inside
+# this factory function, but it would require a redundant
+# call to the decorated function and doesn't look like it would
+# provide any additional value whatsoever.
+def filtered_index_creation_dag_factory(media_type: str):
+    target_alias = f"{media_type}-filtered"
+
+    @task(
+        task_id=f"prevent_concurrency_with_{media_type}_data_refresh",
+    )
+    def prevent_concurrency_with_data_refresh(force: bool, **context):
+        if force:
+            return
+
+        data_refresh_dag_id = f"{media_type}_data_refresh"
+        wait_for_filtered_index_creation = ExternalTaskSensor(
+            task_id="check_for_running_data_refresh",
+            external_dag_id=data_refresh_dag_id,
+            # Set timeout to 0 to prevent retries. If the data refresh DAG is running,
+            # immediately fail the filtered index creation DAG.
+            timeout=0,
+            # Wait for the whole DAG, not just a part of it
+            external_task_id=None,
+            check_existence=False,
+            execution_date_fn=lambda _: get_most_recent_dag_run(data_refresh_dag_id),
+            mode="reschedule",
+        )
+        wait_for_filtered_index_creation.execute(context)
+
+    @task()
+    def generate_index_suffix(default_suffix: str):
+        return default_suffix or uuid.uuid4().hex
+
+    def create_and_populate_filtered_index(
+        origin_index_suffix: str | None,
+        destination_index_suffix: str | None,
+    ):
+        create_payload = {}
+        if origin_index_suffix:
+            create_payload["origin_index_suffix"] = origin_index_suffix
+        if destination_index_suffix:
+            create_payload["destination_index_suffix"] = destination_index_suffix
+
+        return ingestion_server.trigger_and_wait_for_task(
+            action="CREATE_AND_POPULATE_FILTERED_INDEX",
+            model=media_type,
+            data=create_payload or None,
+            timeout=timedelta(days=1),
```

Comment:
I think this is something we might want to set up as a per-media-type configuration. We have the data refresh types defined elsewhere: 

https://github.com/WordPress/openverse/blob/aedc9c16ce5ed11709e2b6f0b42bad77ea1eb19b/catalog/dags/data_refresh/data_refresh_types.py#L59-L72

I notice that you're actually using those types as part of this DAG generator, perhaps we can add a new timeout config option there and leverage that value?

Reply:
Done!